Fails to calculate the type of a generic return type constructed from a Primitive type

Given the following code:
```java
public class GenericClass<A> {

    public A val0;

    public GenericClass(final A value0) {
        this.val0 = value0;
    }

    public static <A> GenericClass<A> with(final A value0) {
        return new GenericClass<A>(value0);
    }

    public static void printGenericClass(GenericClass<Integer> p) {
        System.out.println(p.val0);

    }

    public static void main(String[] args) {
        GenericClass<Integer> p = GenericClass.with(1);
        printGenericClass(p); // OK
        printGenericClass(GenericClass.with(2)); // Unable to calculate the type of a parameter of a method call.
    }
}
```
It crashes on the specified line with the following stacktrace:
```
Exception in thread "main" java.lang.RuntimeException: Unable to calculate the type of a parameter of a method call. Method call: printPair(Pair.with(2)), Parameter: Pair.with(2)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:225)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:240)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:125)
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.resolveDeclaration(JavaSymbolSolver.java:140)
	at com.github.javaparser.ast.expr.MethodCallExpr.resolve(MethodCallExpr.java:313)
        ...
Caused by: java.lang.UnsupportedOperationException: com.github.javaparser.resolution.types.ResolvedPrimitiveType
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.matchTypeParameters(MethodCallExprContext.java:356)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.resolveMethodTypeParameters(MethodCallExprContext.java:345)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethodAsUsage(MethodCallExprContext.java:86)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:584)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:267)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:115)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:448)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:310)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:221)
	... 27 more
```
This is the code segment which throws the original exception:
```java
private void matchTypeParameters(ResolvedType expectedType, ResolvedType actualType, Map<ResolvedTypeParameterDeclaration, ResolvedType> matchedTypeParameters) {
    if (expectedType.isTypeVariable()) {
        if (!actualType.isTypeVariable() && !actualType.isReferenceType()) {
            throw new UnsupportedOperationException(actualType.getClass().getCanonicalName());
        }
        matchedTypeParameters.put(expectedType.asTypeParameter(), actualType);
```
Where, during a debugging session with the code above, actualType is a `PrimitiveTypeUsage{int}` (makes sense) and the expectedType is `TypeVariable`.

We are not sure about how this function is supposed to work, but upon further investigation we think that the issue is that the solver failed to "convert" the int primitive type to Integer, because generics can't use primitives but can indeed use the Integer type (so, the previous bit of code is in fact correct).

We assume this happens with all primitives, but so far we've encountered this problem with both Integer and Double. Testing wether it happens elsewhere shouldn't be too hard.

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/MethodCallExprContext.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/MethodCallExprContext.java
@@ -21,23 +21,6 @@
 
 package com.github.javaparser.symbolsolver.javaparsermodel.contexts;
 
-import com.github.javaparser.ast.expr.Expression;
-import com.github.javaparser.ast.expr.MethodCallExpr;
-import com.github.javaparser.ast.expr.NameExpr;
-import com.github.javaparser.resolution.MethodUsage;
-import com.github.javaparser.resolution.UnsolvedSymbolException;
-import com.github.javaparser.resolution.declarations.*;
-import com.github.javaparser.resolution.types.*;
-import com.github.javaparser.symbolsolver.core.resolution.Context;
-import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
-import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
-import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
-import com.github.javaparser.symbolsolver.model.resolution.Value;
-import com.github.javaparser.symbolsolver.model.typesystem.*;
-import com.github.javaparser.symbolsolver.reflectionmodel.ReflectionClassDeclaration;
-import com.github.javaparser.symbolsolver.resolution.MethodResolutionLogic;
-import com.github.javaparser.utils.Pair;
-
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -45,6 +28,35 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+
+import com.github.javaparser.ast.expr.Expression;
+import com.github.javaparser.ast.expr.MethodCallExpr;
+import com.github.javaparser.ast.expr.NameExpr;
+import com.github.javaparser.resolution.MethodUsage;
+import com.github.javaparser.resolution.UnsolvedSymbolException;
+import com.github.javaparser.resolution.declarations.ResolvedMethodDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedParameterDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedTypeDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedTypeParameterDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedValueDeclaration;
+import com.github.javaparser.resolution.types.ResolvedArrayType;
+import com.github.javaparser.resolution.types.ResolvedLambdaConstraintType;
+import com.github.javaparser.resolution.types.ResolvedReferenceType;
+import com.github.javaparser.resolution.types.ResolvedType;
+import com.github.javaparser.resolution.types.ResolvedTypeVariable;
+import com.github.javaparser.resolution.types.ResolvedUnionType;
+import com.github.javaparser.resolution.types.ResolvedWildcard;
+import com.github.javaparser.symbolsolver.core.resolution.Context;
+import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
+import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.model.resolution.Value;
+import com.github.javaparser.symbolsolver.model.typesystem.ReferenceTypeImpl;
+import com.github.javaparser.symbolsolver.reflectionmodel.MyObjectProvider;
+import com.github.javaparser.symbolsolver.reflectionmodel.ReflectionClassDeclaration;
+import com.github.javaparser.symbolsolver.resolution.MethodResolutionLogic;
+import com.github.javaparser.utils.Pair;
 
 public class MethodCallExprContext extends AbstractJavaParserContext<MethodCallExpr> {
 
@@ -381,10 +393,15 @@ public class MethodCallExprContext extends AbstractJavaParserContext<MethodCallE
 
     private void matchTypeParameters(ResolvedType expectedType, ResolvedType actualType, Map<ResolvedTypeParameterDeclaration, ResolvedType> matchedTypeParameters) {
         if (expectedType.isTypeVariable()) {
-            if (!actualType.isTypeVariable() && !actualType.isReferenceType()) {
-                throw new UnsupportedOperationException(actualType.getClass().getCanonicalName());
+            ResolvedType type = actualType;
+            // in case of primitive type, the expected type must be compared with the boxed type of the actual type
+            if (type.isPrimitive()) { 
+                type = MyObjectProvider.INSTANCE.byName(type.asPrimitive().getBoxTypeQName());
             }
-            matchedTypeParameters.put(expectedType.asTypeParameter(), actualType);
+            if (!type.isTypeVariable() && !type.isReferenceType()) {
+                throw new UnsupportedOperationException(type.getClass().getCanonicalName());
+            }
+            matchedTypeParameters.put(expectedType.asTypeParameter(), type);
         } else if (expectedType.isArray()) {
         	// Issue 2258 : NullType must not fail this search
             if (!(actualType.isArray() || actualType.isNull())) {

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue2083Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue2083Test.java
@@ -1,0 +1,39 @@
+package com.github.javaparser.symbolsolver;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.body.FieldDeclaration;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+
+
+public class Issue2083Test extends AbstractSymbolResolutionTest {
+
+    @Test
+    public void test() {
+        
+        TypeSolver typeSolver = new ReflectionTypeSolver(false);
+        ParserConfiguration config = new ParserConfiguration();
+        config.setSymbolResolver(new JavaSymbolSolver(typeSolver));
+        StaticJavaParser.setConfiguration(config);
+
+        String s = 
+                "import static Simple.SPACES;\n" + 
+                "public class Simple {\n" + 
+                "  public enum IndentType {\n" + 
+                "    SPACES\n" + 
+                "  }\n" + 
+                "  public IndentType c = SPACES;\n" + 
+                "}";
+        CompilationUnit cu = StaticJavaParser.parse(s);
+        FieldDeclaration  fd = cu.findFirst(FieldDeclaration.class).get();
+        assertEquals("Simple.IndentType", fd.resolve().getType().describe());
+        
+    }
+
+}

